### PR TITLE
Support event based smoke test instead of sleep time based to reduce flaky test and faster test

### PR DESCRIPTION
**Pull request by zpoint**
Add some bash functions as global variables:

```python
# cluster
_WAIT_UNTIL_CLUSTER_STATUS_IS
_WAIT_UNTIL_CLUSTER_IS_NOT_FOUND
_WAIT_UNTIL_JOB_STATUS_CONTAINS_MATCHING_JOB_ID 
_WAIT_UNTIL_JOB_STATUS_CONTAINS_WITHOUT_MATCHING_JOB
_WAIT_UNTIL_JOB_STATUS_CONTAINS_MATCHING_JOB_NAME

# managed jobs
_WAIT_UNTIL_MANAGED_JOB_STATUS_CONTAINS_MATCHING_JOB_NAME
```

## Before

```bash
python -m pytest -n 6 tests/test_smoke.py::test_launch_fast_with_autostop tests/test_smoke.py::test_clone_disk_aws tests/test_smoke.py::test_stale_job tests/test_smoke.py::test_aws_stale_job_manual_restart tests/test_smoke.py::test_multi_echo tests/test_smoke.py::test_autostop tests/test_smoke.py::test_managed_jobs_failed_setup --aws

[success] 30.60% tests/test_smoke.py::test_autostop: 1435.7293s
[success] 16.65% tests/test_smoke.py::test_clone_disk_aws: 781.2931s
[success] 12.00% tests/test_smoke.py::test_aws_stale_job_manual_restart: 562.9241s
[success] 11.79% tests/test_smoke.py::test_multi_echo: 553.2864s
[success] 11.28% tests/test_smoke.py::test_managed_jobs_failed_setup: 529.3282s
[success] 11.05% tests/test_smoke.py::test_launch_fast_with_autostop: 518.6133s
[success] 6.62% tests/test_smoke.py::test_stale_job: 310.6923s
7 passed, 1122 warnings in 1437.70s (0:23:57)
```

## After

```bash
python -m pytest -n 6 tests/test_smoke.py::test_launch_fast_with_autostop tests/test_smoke.py::test_clone_disk_aws tests/test_smoke.py::test_stale_job tests/test_smoke.py::test_aws_stale_job_manual_restart tests/test_smoke.py::test_multi_echo tests/test_smoke.py::test_autostop tests/test_smoke.py::test_managed_jobs_failed_setup --aws

[success] 26.90% tests/test_smoke.py::test_autostop: 900.6892s
[success] 19.89% tests/test_smoke.py::test_clone_disk_aws: 666.1263s
[success] 18.24% tests/test_smoke.py::test_multi_echo: 610.6881s
[success] 11.87% tests/test_smoke.py::test_launch_fast_with_autostop: 397.4097s
[success] 9.06% tests/test_smoke.py::test_stale_job: 303.4335s
[success] 7.82% tests/test_smoke.py::test_aws_stale_job_manual_restart: 261.7197s
[success] 6.23% tests/test_smoke.py::test_managed_jobs_failed_setup: 208.6965s
7 passed, 1332 warnings in 902.35s (0:15:02)
```

**40% faster**, and less flaky

And modify some smoke test cases to change to these functions.

The aim is to replace all `sleep xxx` time-based waiting with these event-based waiting functions.

There are over 100 lines of time-based tests that need to be replaced. This PR won't replace all at once(Otherwise the PR would be too large). 
If all goes well, it will be the first PR to offer this possibility, and subsequent PRs will replace the rest.

This should make the test runs faster and less flaky.

image

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: tests/test_smoke.py [hunk 1]
```diff
@@ -1949,7 +2043,8 @@ def test_tpu():
             f'sky logs {name} 1 --status',  # Ensure the job succeeded.
             f'sky launch -y -c {name} examples/tpu/tpu_app.yaml | grep "TPU .* already exists"',  # Ensure sky launch won't create another TPU.
         ],
-        f'sky down -y {name}',
+        'echo "hello"',
+        #f'sky down -y {name}',
```
[the remark this answers is not in the document]
Comment (zpoint) [reply tests/test_smoke.py:2047]: Yes, I forgot to remove the debug code.

File: tests/test_smoke.py [hunk 2]
```diff
@@ -2503,8 +2611,10 @@ def test_autostop(generic_cloud: str):
             f'sky exec {name} echo hi',  # Should restart the timer.
             'sleep 45',
             f's=$(sky status {name} --refresh); echo "$s"; echo; echo; echo "$s"  | grep {name} | grep UP',
-            f'sleep {autostop_timeout}',
-            f's=$(sky status {name} --refresh); echo "$s"; echo; echo; echo "$s"  | grep {name} | grep STOPPED',
+            _WAIT_UNTIL_CLUSTER_STATUS_IS.format(
+                cluster_name=name,
+                cluster_status=ClusterStatus.STOPPED.value,
+                timeout=autostop_timeout),
```
[the remark this answers is not in the document]
Comment (zpoint) [reply tests/test_smoke.py:2617]: Bump up 35 seconds to give 1~3 chances to check.

**Review by romilbhardwaj — APPROVED** [review]
Thanks @zpoint!

**Comment by zpoint** [comment]
Hi @romilbhardwaj, this [PR](https://github.com/zpoint/skypilot/pull/1) adds more test cases and global functions based on current PR.

To avoid disrupting what's already reviewed, I opened another [PR](https://github.com/zpoint/skypilot/pull/1) based on this one. Could you please help review it? Thanks!

**Pull request merged**